### ISSUE TITLE
Required Implementations for Approval

### ISSUE THREAD

**Issue by lindsmith** [ga4gh/product-process#24]
At present, we require two implementations - with APIs, this includes 2 clients and 2 servers. PRC members must be able to run clients and see that they work, and so implementations must be accessible via a URL (supplying access keys as needed). A file interchange format should demonstrate round-trip interchange between the two clients, and an API arrangement should show how the clients can be used to access services provided by the deployments.

There are a number of topics related to required implementations that have not been widely discussed:
- What constitutes an implementation? What if there are implementations that are "in progress" or partially implemented?
- Do implementations need to be accessible? How do we work with implementations that are kept in internal servers? What if providing an access key isn't feasible?
- Do we always need a second implementation? We have heard feedback that a second implementation can be a unnecessary block.
- Related to the above, do we need the two required implementation to show round-trip compatibility? What factors might prevent this from being a feasible exercise?

**Comment by jkbonfield** [comment]
It's clear there are still ambiguities in the CRAM specification, as each time a new implementation arrives (eg the Rust Noodles one) a new tranch of questions arises.  However I believe it's still good in principle to strive for 2 independent implementations.

CRAM 3.1 has really struggled to get a 2nd implementation, so much so that it came from a complete new direction (noodles) before the planned updates to htsjdk.  This is possibly down to complexity.  I suspect CRAM is considerably more complex than most specifications, which likely comes down to both age (it's over 10 year old) and because custom data-compression codecs is inherently a complex task to formalise.  (CRAM is still tiny compared to some MPEG standards for example, but they have the advantage of vastly more resources.)

There is also a more nuanced possibility here.  File format specifications usually describe a data layout and a way to decode the data.  They do not necessarily describe how to encode it.  You can see this is many formal compression specifications, such as deflate (gzip) and Zstd.  It's also true of most ISO file formats.  It can also be seen in internet standards such as HTTP.  Many web browsers can connect to the server and decode the content.   However a web browser is a different beast to a web server.   That means as long as multiple implementations can decode a byte stream, then the decoding side at least has been validated.  

Practically speaking I'm not sure how we address the encode vs decode side.  In an ideal world I would say we also want both multiple encoders and decoders and separating the two is a moot point, so it's worth keeping this in the rules.  There is always some leeway for the GA4GH executive to overrule it should it become necessary to fix blocks (or to allocate resources to ensure the implementation arrives).

**Comment by ohofmann** [comment]
In an ideal world, I continue to see good arguments for having a second, independent implementation. It ensures that the standard specifications can be translated into working, interoperable code and prevents inconsistencies or forks well before a standard is widely adopted. An implementation can still be under development as long as the core functionality related to the standard in question is fully fleshed out.

Given GA4GH's mission I strongly feel that yes, implementations need to be accessible for others to be able to learn from them or to use them side by side with the specs to better understand the standard ideas. For those _one_ reference implementation might be sufficient.

In practice, we run into the issue that groups are not resourced to write reference implementations, and there's little reason to write implementations for a standard that has not yet been released (and which might not be accepted). This leaves us with GA4GH-owned reference implementations that then need to be supported and maintained long-term to not cause issues down the road.

**Comment by ohofmann** [comment]
And maybe a bit more context, I am not sure how to handle the practicalities of reference implementations. 

* it is unclear who should be responsible for the (second) implementation; the standard maintainers do not have the time and resources, but even if they did it defeats the purpose of having an _independent_ piece of evidence showing that the standard works. Driver projects are reluctant to tackle new standard implementations and tend to wait for “the field to catch up"

The current solutions all carry their own risk.

* GA4GH reference implementations can help to get a new standard over the finishing line but then need to be maintained and actively supported. htsget is an example here where this failed, with projects dropping htsget support as our reference server is buggy and endpoints do not work

* Relying on community / volunteer work risks being dependent on even more volunteer time; when community libraries lag behind released standards adoption suffers

**Comment by jkbonfield** [comment]
Some standards bodies view official "reference implementations" as code written for clarity and to demonstrate the specification.  They aren't necessarily intended to be used in anger.  Hence they may not gain lots of users and have a much reduced ongoing maintenance burden.

I don't personally view it as GA4GH's role to be actively maintaining sets of widely used libraries and tool chains, but that doesn't mean it can't do basic implementations to demonstrate feasability and compatibility.  (This is pretty much why I wrote the JavaScript version of the CRAM 3.1 codecs - I don't really intend to be spending time on it.)

It's a tricky one though, and arguably if you've done the effort then is it not better that it's also used?  It depends resource commitments I guess.

**Comment by andrewyatz** [comment]
As one of the original product approval writers I can shine a light on the original intent. We saw at the beginning that a technical standard (APIs primarily) without an implementation would be an odd thing. The two implementation system was put in place in part to force a second implementation that wouldn't have been created by one of the emerging standard writers. Or if it was it hopefully would have happened independent of the first with perhaps one being a reference implementation and another being a real-world example. An early draft even suggested an open source version from memory but this was dropped because of concerns around forcing this. So there's a lot of things mentioned above that certainly have captured the ethos of what was trying to be achieved.

Now onto the modern day.

- We shouldn't reduce the default requirement of x2 implementations and clients at this point in time
- The x2 requirement doesn't prevent a spec going through approval and still be hard/impossible to implement. Plus those are qualitative terms with no basis for the reason why e.g hard to follow docs vs just a hard thing to do
- GA4GH's tech team can provide some level of support for reference implementations but this isn't straightforward as the cost of support needs to be assessed. Those built by the tech team have a better understood cost

Personally though the goal is to ensure we have a good spec and we need to ensure our specification documentation is clear and describe all components of a standard to a sufficient level to allow reproducibility. If we can build frameworks, guidance, tools that allow for that then we could have a different view on the x2 requirement

**Comment by susanfairley** [comment]
Thanks for the comments on this. 

To try to summarise, there are some complicated cases but, in general:

1. Two implementations (ideally including two for encode and two for decode, etc) is favoured
2. At least one of those should be an open reference implementation (although not necessarily used in anger)
3. The second implementation should, ideally, be built independently of the first
4. Developer team implementations could be regarded as reference or second implementations but with the caveat that that team is not an unlimited resource and has to move with GA4GH's priorities (so cannot be guaranteed to provide support)
5. Having a second implementation allows interoperability to be tested as well as clarity of documentation (if a second implementation is built independently)

Writing this, I'm struck that some of these issues also relate to test infrastructure, which also involves the use of documentation and commits to code what the standard means. The comments above reminded me of the work that Cristina did on VCF validation, which I believe helped to clarify the specification.

Happy to hear any further thoughts. I will also review the current wording to aim to check that it reflects these views.

**Comment by jkbonfield** [comment]
Replying to Susan's points above.

1.  We found it much harder to get encode implementations for CRAM.  There is clearly a need for people to process CRAM files, so multiple decoders were written, but very few people are actively writing CRAMs and those that are use existing code so there is simply little driving force behind writing a new encoder.  The only driver infact has been more of a "holy" one - an implementation in a new programming language.

I'd argue that if the specification is a two-way protocol, then it is natural for both encode and decode to be written.  If it is a file format however, the specifications are basically a description of the data layout and instructions for decoding.  Writing a decoder only is 99% of the way to validation of that specification.  So "ideally" seems a good word to use here, but I'm not sure how it applies to actual sign-off.

5.  It's almost impossible to enforce, but ideally we want second implementations that are made by reading the *specification*, and not reading the *code* for an existing implementation.  I am not convinced this always happens though.

**Comment by susanfairley** [comment]
Thanks for the additional comments. I've reviewed the sections relating to implementations and made some additions to try to capture the points discussed.

I have also added an option for, in exceptional circumstances, one implementation being viable (citing the encoder example) and saying reasons need to be presented as to why this is in the community interest. In the revision, it would need PRC approval that this made sense. Happy to hear feedback. Link to the commit is above.
